src/Params/ADnoteParameters.cpp

```diff
@@ -71,7 +71,7 @@ static const Ports voicePorts = {
             if(obj->OscilGn == NULL) return;
         data.obj = obj->OscilGn;
         SNIP
-            OscilGen::realtime_ports.dispatch(msg, data);
+            OscilGen::ports.dispatch(msg, data);
```

Comment:
This dispatch should no longer occur given that the adnoteparameter ports are realtime and this would result in a realtime port invoking a non-realtime port.

Reply:
You mean they should not be dispatched anymore? I just added an `assert(false)` into these ports, but MessageTest failed at `mw->transmitMsg("/presets/copy", "s", "/part0/kit0/adpars/VoicePar0/FMSmp/");`. What could that mean?

Reply:
Oh, PortChecker also failed and SaveOsc aborted... `SaveOsc` probably just aborted because it recurses into all ports, which directly triggers the `assert`. Removing those ports completely fixes it, but then SaveOsc complains that the OscilGen ports were not saved. I'll investigate.

PortChecker may have a similar issue.

Reply:
About the fail of MessageTest: I think `doCopy` in PresetExtractor.cpp assumes that it can do the copy inside the master (read only operation). Maybe we should still allow the OscilGen ports for read-only messages?

Reply:
So, I think PresetExtractor specializes behavior with templated functions, so it should be possible to get it to behave differently in that case. I guess there isn't much harm in leaving these hooks in place though. As it stands the snoop ports should catch everything anyhow before it could make it to the realtime thread. Though actually there is a case where that isn't what would happen...

Automation.
Automation would send messages through the RT side and I think other cases where non-RT parameters are automated, the RT dispatch tree generates a forward() event to send the message to the non-RT thread for processing. So, I guess that should be resolved and possibly tested prior to finishing the branch.

Reply:
Automation: If non-RT messages are really sent to the RT side, then what solutions will we have?

1. Forward *any* non-matched message to MW (sounds very bad, could even lead to message cycling infinitely between MW and Master).
2. Keep those OscilGen non-RT ports in ADnote...
3. Improve the plugin code so it's clever enough to find out which port is RT and which not (Maybe, when the automation is being initialized, dispatching at MW first and check if a reply comes in - if yes, it's a non-RT port, if the data object is forwarded, it's an RT port). That would lead into a direction where Master never forwards anything.

I don't see another solution.

Reply:
OK, we talked about it, the result was to keep the ports and let them only forward, because that seems unavoidable for automation. Resolving.